### ISSUE TITLE
204 No Content should throw error when body is sent instead of parsing it

### ISSUE THREAD

**Issue by neonichu** [apiaryio/dredd#196]
If a 204 HTTP response contains a content-type header, dredd tries to parse the empty response and fails, thus breaking the test. As 204 response must not have a body, this doesn't seem like a good behaviour.

Blueprint:

``` markdown
### Delete a Webhook [DELETE]

+ Request
    + Headers

            Authorization: Bearer b4c0n73n7fu1

+ Response 204 (application/vnd.contentful.management.v1+json)
```

Dredd output:

``` bash
fail: DELETE /spaces/5smsq22uwt4m/webhook_definitions/yolo duration: 673ms
fail: headers: At '/content-type' Missing required property: content-type
body: Expected body: Content-Type is application/vnd.contentful.management.v1+json but body is not a parseable JSON: Parse error on line 1:

^
Expecting 'STRING', 'NUMBER', 'NULL', 'TRUE', 'FALSE', '{', '[', got 'EOF'
body: No validator found for real data media type 'text/plain' and expected data media type 'null'.

request: 
body: 

headers: 
    Authorization:  Bearer XXX
    User-Agent: Dredd/0.5.2 (Darwin 14.4.0; x64)

uri: /spaces/5smsq22uwt4m/webhook_definitions/yolo
method: DELETE


expected: 
headers: 
    Content-Type: application/vnd.contentful.management.v1+json

body: 

statusCode: 204


actual: 
statusCode: 204
headers: 
    accept-ranges: bytes
    access-control-allow-headers: Accept,Accept-Language,Authorization,Cache-Control,Content-Length,Content-Range,Content-Type,DNT,Destination,Expires,If-Match,If-Modified-Since,If-None-Match,Keep-Alive,Last-Modified,Origin,Pragma,Range,User-Agent,X-Http-Method-Override,X-Mx-ReqToken,X-Requested-With,X-Contentful-Version,X-Contentful-Content-Type,X-Contentful-Organization,X-Contentful-Skip-Transformation
    access-control-allow-methods: DELETE,GET,HEAD,POST,PUT,OPTIONS
    access-control-allow-origin: *
    access-control-expose-headers: Etag
    access-control-max-age: 1728000
    cache-control: max-age=0
    date: Thu, 21 May 2015 12:13:03 GMT
    server: nginx
    status: 204 No Content
    x-contentful-request-id: c0e-1770848933
    connection: keep-alive

body: 
```

**Comment by netmilk** [comment]
Hi @neonichu,

thanks for the report and for opening this issue and for the feedback! I'm excited you're using Dredd! 

I'm sorry, but I'm afraid this is not a bug. My intention here was when you declare body should be a parseable JSON in the `content-type` header, Dredd tries to parse it regardless the status code `204 No Content`. 

My opinion is that it makes no sense to send any `content-type` header, because there is no content. :)

See [RFC2616, §7.2.1]http://tools.ietf.org/search/rfc2616#section-7.2.1()

> Any HTTP/1.1 message containing an entity-body SHOULD include a Content-Type header field defining the media type of that body.

**Comment by neonichu** [comment]
Dredd is awesome, thanks a lot for making it :+1: 

I guess this is an edge case :) 

Section "10.2.5 204 No Content" of the same RFC states:

>The 204 response MUST NOT include a message-body, and thus is always terminated by the first empty line after the header fields.

Feel free to close this, though, if this works as you intended.

**Comment by netmilk** [comment]
Thanks for pointing this out. Do you think that Dredd should not parse body on `204` status code and throw an error if the body is present?

**Comment by neonichu** [comment]
I think so, as "MUST NOT" is defined as "absolute prohibition" in RFCs, sending a body with a `204` response is clearly an error.

**Comment by netmilk** [comment]
Well, I'll rename this issue if you don't mind :) Thank you!

**Comment by milch** [comment]
I ran into a similar issue, because I am using `rails` for my API. `rails` is using `rack`, which is always emitting the `Content-Type` header, even if there is no body. To solve this, I put this script I found on [StackOverflow](http://stackoverflow.com/questions/21855766/disable-content-type-header-in-rails-rack) (and adapted it) into my `application.rb`:

```ruby
class RemoveContentType
  def initialize(app)
    @app = app
  end

  def call(env)
    status, headers, body = @app.call(env)
    bbody = body.body

    #Remove Content-Type if there is no body
    headers.delete('Content-Type') if bbody.length == 0
    return [status, headers, body] 
  end
end

```

And then called `config.middleware.use "::RemoveContentType"` in the config part of the same file.

**Comment by netmilk** [comment]
@milch Thanks! :candy:

**Comment by ungrim97** [comment]
I am hitting this issue as well. Returning from an OPTIONS request is failing in Dredd as it attempts to parse the JSON when there is no content.

I would argue that there doesn't have to be a Body if there is a Content-Type header, but there must be a Content-type header if there is a body.

Certainly the RFC indicates that in the event of a 204 then the Body should be ignored.

**Comment by netmilk** [comment]
Just letting know I proposed a solution for this issue here: #556

**Comment by honzajavorek** [comment]
Hi folks, please check out a new proposed behavior and let me know if that would work for you: https://github.com/apiaryio/dredd/issues/556#issuecomment-252206782

**Comment by honzajavorek** [comment]
I believe this is already addressed by [v4.7.1](https://github.com/apiaryio/dredd/releases/tag/v4.7.1) and further, I just forgot to close this one. Thanks everyone involved and I'm sorry it took so long to fix this.
